internal/indexer/indexer.go

```diff
@@ -732,3 +937,93 @@ func (i *Indexer) Stats() IndexerStats {
 		NumElements: i.emitter.NumElements(),
 	}
 }
+
+func (i *Indexer) setRangeForPosition(position token.Position, id uint64) {
+	i.stripedMutex.LockKey(position.Filename)
+	i.ranges[position.Filename][position.Offset] = id
+	i.stripedMutex.UnlockKey(position.Filename)
+}
+
+// findBestPackageDefinitionPath searches paths in possiblePaths and finds the one that seems best.
+// Chooses one with documentation if possible, otherwise looks for most similar name.
+func findBestPackageDefinitionPath(packageName string, possiblePaths []DeclInfo) (string, error) {
+	if len(possiblePaths) == 0 {
+		return "", errors.New("must have at least one possible path")
+	}
+
+	pathsWithDocs := []DeclInfo{}
+	for _, v := range possiblePaths {
+		if v.HasDoc {
+			pathsWithDocs = append(pathsWithDocs, v)
+		}
+	}
+
+	// The idiomatic way is to _only_ have one .go file per package that has a docstring
+	// for the package. This should generally return here.
+	if len(pathsWithDocs) == 1 {
+		return pathsWithDocs[0].Path, nil
+	}
+
+	// If we for some reason have more than one .go file per package that has a docstring,
+	// only consider returning paths that contain the docstring (instead of any of the possible
+	// paths).
+	if len(pathsWithDocs) > 1 {
+		possiblePaths = pathsWithDocs
+	}
+
+	// Try to only pick non _test files for non _test packages and vice versa.
+	possiblePaths = filterBasedOnTestFiles(possiblePaths, packageName)
+
+	// Find the best remaining path.
+	// Chooses:
+	//     1. doc.go
+	//     2. exact match
+	//     3. computes levenshtein and picks best score
+	minDistance, bestPath := math.MaxInt32, ""
+	for _, v := range possiblePaths {
+		fileName := fileNameWithoutExtension(v.Path)
+
+		if "doc.go" == path.Base(v.Path) {
+			return v.Path, nil
+		}
+
+		if packageName == fileName {
+			return v.Path, nil
+		}
+
+		distance := levenshtein.ComputeDistance(packageName, fileName)
+		if distance < minDistance {
+			minDistance = distance
+			bestPath = v.Path
+		}
+	}
+
+	return bestPath, nil
+}
+
+func fileNameWithoutExtension(fileName string) string {
+	return strings.TrimSuffix(fileName, path.Ext(fileName))
+}
+
+func filterBasedOnTestFiles(possiblePaths []DeclInfo, packageName string) []DeclInfo {
+	preferredPaths := []DeclInfo{}
+	if !strings.HasSuffix(packageName, "_test") {
+		for _, v := range possiblePaths {
+			if !strings.HasSuffix(v.Path, "_test.go") {
```

Comment:
You could collapse these loops with the condition:

`if strings.HasSuffix(packageName, "_test") == strings.HasSuffix(v.Path, "_test.go")`